src/mapgen/mapgen.h

```diff
@@ -76,29 +76,41 @@ enum GenNotifyType {
 	GENNOTIFY_LARGECAVE_BEGIN,
 	GENNOTIFY_LARGECAVE_END,
 	GENNOTIFY_DECORATION,
+	GENNOTIFY_CUSTOM, // user-defined data
 	NUM_GENNOTIFY_TYPES
 };
 
-struct GenNotifyEvent {
-	GenNotifyType type;
-	v3s16 pos;
-	u32 id;
-};
-
 class GenerateNotifier {
 public:
+	struct GenNotifyEvent {
+		GenNotifyType type;
+		v3s16 pos;
+		u32 id; // for GENNOTIFY_DECORATION
+	};
+
 	// Use only for temporary Mapgen objects with no map generation!
 	GenerateNotifier() = default;
-	GenerateNotifier(u32 notify_on, const std::set<u32> *notify_on_deco_ids);
-
-	bool addEvent(GenNotifyType type, v3s16 pos, u32 id=0);
-	void getEvents(std::map<std::string, std::vector<v3s16> > &event_map);
+	// normal constructor
+	GenerateNotifier(u32 notify_on, const std::set<u32> *notify_on_deco_ids,
+		const std::set<std::string> *notify_on_custom);
+
+	bool addEvent(GenNotifyType type, v3s16 pos);
+	bool addDecorationEvent(v3s16 pos, u32 deco_id);
+	bool setCustom(const std::string &key, const std::string &value);
+	void getEvents(std::map<std::string, std::vector<v3s16>> &map) const;
+	const StringMap &getCustom() const { return m_notify_custom; }
```

Comment:
Unclear name because `custom` alone is too generic. Suggestion: `getCustomGenNotifyData` (or shorter, if you have a better idea).